### PR TITLE
Composite setting key for QAction & QShortcuts keyboard shortcuts

### LINKED ISSUE #39903: Keyboard shortcut for "Save As..." (project) and "Save As..." (export layer) keeps overwriting themselves
I would like to set keyboard shortcut for **"Save As..."** in Project menu and in Layer menu differently !
Currently they keep overwriting themselvs (after restart).
bug_qgis_saveas_keyboardshortcut_overwriting

### LINKED ISSUE #51856: Shortcut configurations are language-sensitive
### What is the bug or the crash?

QGIS shortcut manager seems to use the translated text to save/load shortcuts to/from the profile settings, causing issues for example when providing defaults via global settings files for new profiles or when changing the language after manually settings shortcuts.

Could the shortcut manager use object names for saving, and support reading of both translated texts and object names (as a fallback when there is no match for the text), if that could be backwards-compatible enough to qualify for 3.28 LTR backporting?

### Steps to reproduce the issue

- Open QGIS with a fresh profile (opens with your system language)
- Set some shortcut from the manager GUI, it is saved to profile with the translated text
- Change QGIS language to something else and restart with the same profile
- Observe that the shortcut is no longer set but instead uses the default
- Reapply the same shortcut on the newly-set language, a new "duplicate" line appears in the .ini file

image

image

### Versions

QGIS 3.22.16 (from OSGeo4W network installer) on Windows 10

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by YoannQDQ**
Fixes #39903
Fixes #51856

## Description

`QgsShortcutManager` keeps track of registered `QAction`s by action text.
This causes 2 problems:

- Shortcuts don't work anymore if the locale is changed, and exported shortcuts file cannot be used across locales
- If two action share the same text, user can set a shortcut for each action, but when QQGIS is restarted, they will both have the same value (see #39903), which causes an ambiguous shortcut by the way.

This PR introduces the following changes:
- Use `objectName` (if not empty) instead of the action text
- Add a `section` parameter in register methods in `QgsShortcutManager` to allow disambiguation. Use this by default when all actions and shortcuts of an item are registered recursively. For each registered `QAction`, the matching setting key will be `/shortcuts/parent/parent2/objectName`
- Bump the version of exported shortcut files to 1.1. Old files can still be loaded. New files (1.1) contains the setting key instead of the action text, and can be loaded even if they were saved in a different locale.

### CODE REVIEW

**Comment by YoannQDQ** [comment]
spell_check fails because disambiguate is not a word? Isn't it though?

File: python/gui/auto_generated/qgsshortcutsmanager.sip.in [hunk 1]
```diff
@@ -291,6 +296,22 @@ Returns a shortcut by its name, or ``None`` if nothing found
...
+%Docstring
+Returns the full settings key matching the QShortcut or QAction
+Return an empty QString if the QObject is not registered
+
+.. versionadded:: 3.30
+%End
+
+    QObject *objectForSettingKey( const QString &name ) const;
+%Docstring
+Returns the QShortcut or QAction matching the the full setiing key matching
```
Comment (DelazJ) [inline python/gui/auto_generated/qgsshortcutsmanager.sip.in:311]:
```suggestion
Returns the QShortcut or QAction matching the full setting key
```
? or "... matching name"?

Comment (YoannQDQ) [reply python/gui/auto_generated/qgsshortcutsmanager.sip.in]: Oops... My bad.

File: python/gui/auto_generated/qgsshortcutsmanager.sip.in [hunk 2]
```diff
@@ -36,25 +36,27 @@ Constructor for QgsShortcutsManager.
...
 
-    void registerAllChildren( QObject *object, bool recursive = false );
+    void registerAllChildren( QObject *object, bool recursive = false, const QString &section = QString() );
 %Docstring
 Automatically registers all QActions and QShortcuts which are children of the
 passed object.
 
 :param object: parent object containing actions and shortcuts to register
 :param recursive: set to ``True`` to recursively add child actions and shortcuts
+:param section: since QGIS 3.30 Disambiguate shortcuts with the same objectName
```
Comment (nyalldawson) [inline python/gui/auto_generated/qgsshortcutsmanager.sip.in:46]: Can you update these to 3.32?

**Comment by nyalldawson** [comment]
@YoannQDQ sorry, i had a pending review here which I never hit submit for! :open_mouth:

File: python/gui/auto_generated/qgsshortcutsmanager.sip.in (hunk 2 above)
Comment (YoannQDQ) [reply python/gui/auto_generated/qgsshortcutsmanager.sip.in]: Done!

File: src/gui/qgsshortcutsmanager.cpp [hunk 3]
```diff
@@ -22,92 +22,83 @@
 #include <QWidgetAction>
 
 QgsShortcutsManager::QgsShortcutsManager( QObject *parent, const QString &settingsRoot )
-  : QObject( parent )
-  , mSettingsPath( settingsRoot )
+  : QObject( parent ), mSettingsPath( settingsRoot )
```
[the remark this answers is not in the document]
Comment (YoannQDQ) [reply src/gui/qgsshortcutsmanager.cpp]: Ok

File: src/gui/qgsconfigureshortcutsdialog.cpp [hunk 4]
```diff
@@ -262,22 +269,44 @@ void QgsConfigureShortcutsDialog::loadShortcuts()
...
   }
 
+  const QString version = root.attribute( QStringLiteral( "version" ) );
+
   if ( root.attribute( QStringLiteral( "locale" ) ) != currentLocale )
   {
-    QMessageBox::information( this, tr( "Loading Shortcuts" ),
-                              tr( "The file contains shortcuts created with different locale, so you can't use it." ) );
-    return;
+    if ( version < "1.1" )
```
[the remark this answers is not in the document]
Comment (YoannQDQ) [reply src/gui/qgsconfigureshortcutsdialog.cpp]: Done

File: src/gui/qgsconfigureshortcutsdialog.cpp [hunk 5]
```diff
@@ -89,15 +90,16 @@ void QgsConfigureShortcutsDialog::populateActions()
...
-    if ( QAction *action = qobject_cast< QAction * >( obj ) )
+    if ( QAction *action = qobject_cast<QAction *>( obj ) )
     {
       actionText = action->text();
       actionText.remove( '&' ); // remove the accelerator
       sequence = action->shortcut().toString( QKeySequence::NativeText );
       icon = action->icon();
     }
-    else if ( QShortcut *shortcut = qobject_cast< QShortcut * >( obj ) )
+    else if ( QShortcut *shortcut = qobject_cast<QShortcut *>( obj ) )
```
[the remark this answers is not in the document]
Comment (YoannQDQ) [reply src/gui/qgsconfigureshortcutsdialog.cpp]: Ok

File: src/gui/qgsconfigureshortcutsdialog.cpp [hunk 6]
```diff
@@ -39,20 +39,21 @@
 #include <QTextCharFormat>
 
 QgsConfigureShortcutsDialog::QgsConfigureShortcutsDialog( QWidget *parent, QgsShortcutsManager *manager )
-  : QDialog( parent )
-  , mManager( manager )
+  : QDialog( parent ), mManager( manager )
```
[the remark this answers is not in the document]
Comment (YoannQDQ) [reply src/gui/qgsconfigureshortcutsdialog.cpp]: Ok

File: src/gui/qgsconfigureshortcutsdialog.cpp [hunk 7]
```diff
@@ -39,20 +39,21 @@
...
 {
   setupUi( this );
   QgsGui::enableAutoGeometryRestore( this );
 
   mSaveMenu = new QMenu( this );
   mSaveUserShortcuts = new QAction( tr( "Save User Shortcuts…" ), this );
   mSaveMenu->addAction( mSaveUserShortcuts );
-  connect( mSaveUserShortcuts, &QAction::triggered, this, [this] { saveShortcuts( false ); } );
+  connect( mSaveUserShortcuts, &QAction::triggered, this, [this]
+  { saveShortcuts( false ); } );
```
[the remark this answers is not in the document]
Comment (YoannQDQ) [reply src/gui/qgsconfigureshortcutsdialog.cpp]: ok

File: src/gui/qgsshortcutsmanager.h [hunk 8]
```diff
@@ -54,53 +54,58 @@ class GUI_EXPORT QgsShortcutsManager : public QObject
      * passed object.
      * \param object parent object containing actions and shortcuts to register
      * \param recursive set to TRUE to recursively add child actions and shortcuts
+     * \param section since QGIS 3.30 Allow to disambiguate shortcuts with the same objectName
```
[the remark this answers is not in the document]
Comment (YoannQDQ) [reply src/gui/qgsshortcutsmanager.h]: ok

File: src/gui/qgsconfigureshortcutsdialog.cpp [hunk 9]
```diff
@@ -262,22 +269,44 @@ void QgsConfigureShortcutsDialog::loadShortcuts()
...
   QString actionShortcut;
+  QString actionSettingKey;
 
   QDomElement child = root.firstChildElement();
   while ( !child.isNull() )
   {
-    actionName = child.attribute( QStringLiteral( "name" ) );
     actionShortcut = child.attribute( QStringLiteral( "shortcut" ) );
-    mManager->setKeySequence( actionName, actionShortcut );
+    if ( version < "1.1" )
```
Comment (nyalldawson) [inline src/gui/qgsconfigureshortcutsdialog.cpp:297]: This should use QgsProjectVersion for the comparison too

Comment (YoannQDQ) [reply src/gui/qgsconfigureshortcutsdialog.cpp]: My bad. Done.

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**